[ACIX-894] Refactor gitlab linter tasks logic

Ishirui opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?

Cleans up and unifies the way gitlabci linting tasks run, especially in regards to how they generate the "full gitlab config". 
Also fixes some wrong behavior with `gitlab_change_paths` and `job_change_path`, these two tasks were linting only the 'root' pipeline (defined in `.gitlab-ci.yml`), but not any "called pipelines", like `.gitlab/deploy_containers`

**Call graph before the PR:**
![Gitlab CI Linters Call Graph 0 Explained](https://github.com/user-attachments/assets/76119cf1-e5b4-4bd1-9f08-518058ec6822)
> Key:
> - **Grey arrows:** Calls
> - **Purple arrows:** Need file as input
> - **Blue:** Functions handling the "full gitlab config", i.e. with multiple entrypoints, i.e. including "called pipelines"
> - **Red:** Functions handling only a single config object, i.e. only one entrypoint (usually the main `.gitlab-ci.yml`)

**Call graph after the PR:**
![Gitlab CI Linters Call Graph 3 Raw](https://github.com/user-attachments/assets/968ce1cd-231e-4df3-9a99-61fa3d55c1ef)
Notable changes:
- Fixed `job_change_path` and `gitlab_change_paths` to lint all entrypoints instead of only the root `.gitlab-ci.yml`
- Removed `get_gitlab_ci_configuration` and `generate_gitlab_full_configuration`, instead making the tasks call `get_all_gitlab_ci_configurations` directly. These functions had too little useful functionality and made the code much harder to read.
- Renamed the recursive helper function `get_ci_configurations` to `_traverse_config_search_triggers`
- `test_gitlab_configuration` no longer calls methods leading to `resolve_gitlab_ci_configurations`, instead taking the config object as input. This avoids calling the `/lint` endpoint one extra time, improving efficiency. It also does not depend on other methods in the call graph anymore, so is not represented above.
- Made `get_gitlab_ci_lintable_jobs` take config files as input only optionally, falling back to generating the config objects if necessary

### Motivation

Before trying to implement a new top-level task for running all gitlabci-related linters [ACIX-470], I needed to make sure that different linting tasks would not try to regenerate the unified gitlab config multiple times, to ensure these tasks would run as quickly as possible.

But due to the organic nature of how these tasks were built over time, there were many different paths to generating this config, and many different functions all with the same goal.
With a more unified method of generating this config, we make the code more readable and efficient, while assuring no task tries to regenerate the config if it does not need to.

### Describe how you validated your changes
<!--
Validate your changes before merge, ensuring that:
* Your PR is tested by static / unit / integrations / e2e tests
* Your PR description details which e2e tests cover your changes, if any
* The PR description contains details of how you validated your changes. If you validated changes manually and not through automated tests, add context on why automated tests did not fit your changes validation.

If you want additional validation by a second person, you can ask reviewers to do it. Describe how to set up an environment for manual tests in the PR description. Manual validation is expected to happen on every commit before merge.

Any manual validation step should then map to an automated test. Manual validation should not substitute automation, minus exceptions not supported by test tooling yet.
-->

- All affected tasks run without errors on the current gitlab config
- Inputting a broken gitlab config (simple change for each of the linter tasks) does result in an error
- Some tasks are already covered by unit tests

### Possible Drawbacks / Trade-offs

- Performance impact: By fixing the two tasks mentioned above, they become a bit slower as they have to handle multiple entrypoints instead of only the root one (should be very minimal)
- Behavioral change: To the best of my knowledge, there should be none, but I might have missed some transformation in one of the helper functions that is not being executed anymore
- Loss of habit: As with any significant refactor, anyone familiar with the old code might not recognize it. In my opinion it is more easily understandable and readable in its current state, however.

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

[ACIX-470]: https://datadoghq.atlassian.net/browse/ACIX-470?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Ishirui committed:
refactor(gitlab_api): Remove references to `generate_gitlab_full_configuration`

This function was used in very similar contexts and for similar purposes as `get_gitlab_ci_configuration`, we try to consolidate both into a single function
Note that the new function is slower than the original, but generates a more "complete" gitlab config - the old config objects' `.keys()` were subsets of the new objects'.
I will change the usages of this new function, adding flags to match the speed of the original.

Ishirui committed:
refactor(tasks): Merge `input_file` and `input_config` args

Ishirui committed:
refactor(gitlab_api): Never return yaml dump of config objects

Always manipulate native python dicts instead of yaml-encoded strings. The dump only happens when needed.

Ishirui committed:
fix(tasks): Make `gitlab_change_paths` and `job_change_path` handle multiple entrypoints/configs

In the repo we actually have multiple independent gitlab configs, that are coupled via trigger jobs (`trigger:`). However, they are completely independent from a syntax and linting perspective, and have different "entry points" ("main files", like the usual `.gitlab-ci.yml`). These two tasks used to only lint the main pipeline, but now also lint any triggered pipelines that might exist in the repo.

Ishirui committed:
refactor(gitlab_api): Move functions around and improve docstrings

Ishirui committed:
refactor(tasks): Rename `with_lint` argument to `partial_resolve`

This argument was named this way because it controls whether `resolve_gitlab_ci_configuration` calls the `/lint` endpoint - but that endpoint actually does more important things than just lint the config file. It is actually mostly used for resolving `extend` and `!reference` tags in the configs, which is almost always what we want.

Ishirui committed:
refactor(gitlab_api): Create `post_process_gitlab_ci_configuration` and remove `get_gitlab_ci_configuration` method

After splitting post-processing responsibilities into its own reusable function (instead of each function using some of the sub-methods, and never in the same way !), `get_gitlab_ci_configuration` has too few responsibilities and can be removed

Ishirui committed:
fix(gitlab_api): Remove extraneous calls to `/lint` endpoint

Instead of doing a partial resolve in `gitlab_ci`, then resolving in `test_gitlab_configuration`, we resolve fully once at the beginning.
Changing the context, as is done in each call to `test_gitlab_configuration`, should not affect how the configuration gets resolved (it simply changes the values of variables in the CI config)

Ishirui committed:
refactor(gitlab_api): Use `post_process_gitlab_ci_configuration` for postprocessing in `get_all_gitlab_ci_configurations`

Ishirui committed:
feat(linter): Make passing gitlabci diff/config files optional

Three linting tasks `gitlab_ci_shellcheck`, `gitlab_change_paths` and `gitlab_ci_jobs_needs_rules` required passing a config/diff file as argument, as generated by `compute_gitlab_ci_config`.
This is now optional, if not passed the contents of the file will be generated on-the-fly with `get_all_gitlab_ci_configurations`

CelianR commented on tasks/gitlab_helpers.py:
```diff
@@ -252,7 +252,7 @@ def print_ci(
     expand_matrix: bool = False,
     git_ref: str | None = None,
-    with_lint: bool = True,
+    partial_resolve: bool = True,
```
NIT: Maybe `no_lint` or `resolve_only_includes` would be more explicit?

Ishirui replied:
I like `resolve_only_includes` - a bit longer but definitely clearer :+1:

CelianR commented on tasks/libs/linter/gitlab.py:
```diff
@@ -64,34 +65,45 @@ def list_get_parameter_calls(file):
+        A (jobs, full_config) tuple.
+        `jobs` is itself a tuple of (job_name: str, job_contents: dict). If `only_names` is True, it will be a simple list of all the job names.
+        `full_config` is a gitlabci config object, of the same structure as returned by `get_all_gitlab_ci_configurations`
+
```
```suggestion
```

CelianR commented on tasks/libs/linter/gitlab.py:
```diff
@@ -64,34 +65,45 @@ def list_get_parameter_calls(file):
             diff = MultiGitlabCIDiff.from_dict(yaml.safe_load(f))
 
-        full_config = diff.after
+        configs = diff.after  # type: ignore
```
NIT: Do we really want to add `type: ignore`? Maybe it's better to not type `configs`

CelianR replied:
Most of the time we don't use this but as you wish, I think adding an explicit check is too much for this

CelianR reviewed (approved):
LGTM if the linters are passing 👍

Ishirui committed:
refactor(gitlab_api): Rename `partial_resolve` arg to `resolve_only_includes`

[Review comment](https://github.com/DataDog/datadog-agent/pull/37417#discussion_r2125925396) from @CelianR

Ishirui committed:
fix(linter): Use a minimally-resolved config object in `gitlab_ci`

This is done for performance reasons, and is the way it is currently done on `main`, but I accidentally changed during the refactor.
Without this change the `dda inv gitlab-ci -t main` invocation takes ~60s, up from ~40s as it is currently on `main`. This is due to the extra time taken by creating a fully-resolved gitlabci config object instead of the actually-needed minimal one.
This commit makes sure we use a minimally-resolved one, restoring the performance currently on main.

Ishirui committed:
fix(gitlab_api): Always resolve `include`s when running `resolve_gitlab_ci_configurations`

Ishirui committed:
fix(linter): Remove postprocessing clean and filtering from `test_gitlab_configuration`

The config object passed to `test_gitlab_configuration` is minimally-resolved (only `include`s are resolved), for performance reasons.
Before testing the validity of the config object, we should not be removing any dotted jobs, `extends` tags or the like, otherwise the configuration will be considered invalid.
Thus we need to disable the `clean` and `do_filtering` options from the post-processing function, as that is exactly what they do.

Ishirui committed:
fix(gitlab_api): Deepcopy config object when applying post-processing

Otherwise there can be crosstalk between different config objects, for example when applying different `context` sets of variables in `gitlab_ci`.
